runtime/src/accounts.rs

```diff
@@ -451,39 +516,70 @@ impl Accounts {
-        if res[i].is_err() || raccs.is_err() {
-            continue;
+    /// Commit remaining credit-only changes, regardless of reference count
+    pub fn commit_credits(&self, ancestors: &HashMap<Fork, usize>, fork: Fork) {
+        let credit_only_account_locks = self.credit_only_account_locks.read().unwrap();
+        for (pubkey, credit_only_lock) in credit_only_account_locks.iter() {
+            let credit = credit_only_lock.credits.load(Ordering::Relaxed);
+            if credit > 0 {
+                let mut account = self
+                    .load_slow(ancestors, pubkey)
+                    .map(|(account, _)| account)
+                    .unwrap_or_default();
+                account.lamports += credit;
+                self.store_slow(fork, pubkey, &account);
+            }
         }
+        drop(credit_only_account_locks);
+        self.clear_credit_only_account_locks();
+    }
 
-        let message = &txs[i].message();
-        let acc = raccs.as_ref().unwrap();
-        for ((key, account), credit) in message
-            .account_keys
-            .iter()
-            .zip(acc.0.iter())
-            .zip(acc.2.iter())
-        {
-            if !accounts.contains_key(key) {
-                accounts.insert(key, (account, 0));
-            } else if *credit > 0 {
-                // Credit-only accounts may be referenced by multiple transactions
-                // Collect credits to update account lamport balance before store.
-                if let Some((_, c)) = accounts.get_mut(key) {
-                    *c += credit;
+    pub fn clear_credit_only_account_locks(&self) {
+        let mut credit_only_account_locks = self.credit_only_account_locks.write().unwrap();
+        credit_only_account_locks.clear();
+    }
+
+    fn collect_accounts<'a>(
+        &self,
+        txs: &'a [Transaction],
+        res: &'a [Result<()>],
+        loaded: &'a mut [Result<(InstructionAccounts, InstructionLoaders, InstructionCredits)>],
+    ) -> HashMap<&'a Pubkey, (&'a Account, bool)> {
+        let mut accounts: HashMap<&Pubkey, (&Account, bool)> = HashMap::new();
+        for (i, raccs) in loaded.iter_mut().enumerate() {
+            if res[i].is_err() || raccs.is_err() {
+                continue;
+            }
+
+            let message = &txs[i].message();
+            let acc = raccs.as_mut().unwrap();
+            for (((i, key), account), credit) in message
+                .account_keys
+                .iter()
+                .enumerate()
+                .zip(acc.0.iter())
+                .zip(acc.2.iter())
+            {
+                if !accounts.contains_key(key) {
+                    accounts.insert(key, (account, message.is_debitable(i)));
+                }
+                if *credit > 0 {
+                    // Increment credit-only account balance Atomic
+                    if accounts.get_mut(key).is_some() {
+                        self.credit_only_account_locks
+                            .write()
```

Comment:
This can be credit_only_locks.get(**k).unwrap(), the entry has to exist by this point or there's something wrong